### ISSUE TITLE
mocha -r esm: ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING

### ISSUE THREAD

**Issue by Yonom** [mochajs/mocha#4671]
### Prerequisites

- [x] Checked that your issue hasn't already been filed by cross-referencing [issues with the `faq` label](https://github.com/mochajs/mocha/issues?utf8=%E2%9C%93&q=is%3Aissue%20label%3Afaq%20)
- [x] Checked next-gen ES issues and syntax problems by using the same environment and/or transpiler configuration without Mocha to ensure it isn't just a feature that actually isn't supported in the environment in question or a bug in your code.
- [x] 'Smoke tested' the code to be tested by running it outside the real test suite to get a better sense of whether the problem is in the code under test, your usage of Mocha, or Mocha itself
- [x] Ensured that there is no discrepancy between the locally and globally installed versions of Mocha. You can find them with: `node node_modules/.bin/mocha --version`(Local) and `mocha --version`(Global). We recommend that you _not_ install Mocha globally.

### Description

Since version v9.0.0 mocha crashes when I try to use it with ESM.

### Steps to Reproduce

**Expected behavior:** [What you expect to happen]
Mocha should run successfully.

**Actual behavior:** [What actually happens]

Mocha crashes.

Stack trace:
```js
TypeError [ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING]: A dynamic import callback was not specified.
    at new NodeError (node:internal/errors:363:5)
    at exports.importModuleDynamicallyCallback (node:internal/process/esm_loader:34:9)
    at formattedImport (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/mocha/lib/esm-utils.js:7:31)
    at Object.exports.requireOrImport (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/mocha/lib/esm-utils.js:48:38)
    at Object.exports.loadFilesAsync (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/mocha/lib/esm-utils.js:73:34)
    at Mocha.loadFilesAsync (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/mocha/lib/mocha.js:467:19)
    at singleRun (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/mocha/lib/cli/run-helpers.js:125:15)
    at exports.runMocha (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/mocha/lib/cli/run-helpers.js:190:10)
    at Object.exports.handler (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/mocha/lib/cli/run.js:366:11)
    at /Users/yonom/GitHub/test/mocha-bug-report/node_modules/yargs/build/index.cjs:443:71
    at process.runNextTicks [as _tickCallback] (node:internal/process/task_queues:61:5)
    at /Users/yonom/GitHub/test/mocha-bug-report/node_modules/esm/esm.js:1:34535
    at /Users/yonom/GitHub/test/mocha-bug-report/node_modules/esm/esm.js:1:34176
    at process.<anonymous> (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/esm/esm.js:1:34506)
    at Function.<anonymous> (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/esm/esm.js:1:296856)
    at Function.<anonymous> (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/esm/esm.js:1:296555)
    at Function.<anonymous> (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/esm/esm.js:1:284879)
    at Object.apply (/Users/yonom/GitHub/test/mocha-bug-report/node_modules/esm/esm.js:1:199341)
    at node:internal/main/run_main_module:17:47
```

**Reproduces how often:** [What percentage of the time does it reproduce?]

100% (always)

### Versions

- The output of `mocha --version` and `node node_modules/.bin/mocha --version`: 9.0.1
- The output of `node --version`: v16.4.0 (also tested on v14.17.1)
- Your operating system
  - name and version: macOS 11.4 (also tested on GitHub Actions Ubuntu 20.04)
  - architecture (32 or 64-bit): 64-bit
- Your shell (e.g., bash, zsh, PowerShell, cmd): zsh (also tested on bash)
- Your browser and version (if running browser tests): N/A
- Any third-party Mocha-related modules (and their versions): N/A
- Any code transpiler (e.g., TypeScript, CoffeeScript, Babel) being used (and its version): ESM v3.2.25

### Additional Information

Repository for reproduction: https://github.com/Yonom/mocha-bug-report

You can run ```yarn test``` in the (mostly empty) repo abobe to see the error occur. No actual tests are needed, an empty test file suffices to trigger the issue (but the error still happens even with real tests existing of course).

**Comment by juergba** [comment]
@Yonom `esm` is a package which hasn't been maintained for more than two years and IMO shouldn't be required anymore. Node's native ESM implementation is stable and should be the solution of choice. It's obvious that you can't mix both solution at the same time.

Try to pass `-r esm` directly to Node: `node -r esm node_modules/mocha/bin/mocha`. Probably will fail as well.
From my side we will not offer any fix for this issue.
@giltayar 

PS: your reproduction test is empty.

**Comment by Yonom** [comment]
@juergba Thank you for your answer. I have added the reproduction steps in the question now. Sorry if that was confusing.

You can run ```yarn test``` in the (mostly empty) repo to see the error occur. No actual tests are needed, an empty file suffices to trigger the issue.

Is `esm` deprecated? Because I do not see any signs of it being so. The package still has 1.6M+ weekly downloads.

**Comment by juergba** [comment]
`esm` is abandoned. No new versions since two years, no answers to issues, the test matrix is Node versions [8, 10, 11, 12] (most of them end-of-life) and no, officially it's not deprecated.

**Comment by giltayar** [comment]
@Yonom Thanks for the repro. It reproduced the problem and enabled me to verify that the fix is REALLY easy, so I'll probably do it in the next few days. In the meantime, feel free to downgrade to v8.

As for `esm`, I believe @juergba is right. JDD (who wrote the package) probably abandoned it once native ESM in Node.js was mature enough. I would strongly advise you to start migrating your codebases away from this and towards native ESM, as the the ecosystem is (pretty rapidly I might say) adopting native ESM and abandoning what are now called "faux modules".

Shameless plug: my ESM migration guide: https://gils-blog.tayar.org/posts/using-jsm-esm-in-nodejs-a-practical-guide-part-1/ 😊

**Comment by giltayar** [comment]
@juergba I know you added `wontfix` to the label, but it's a VERY easy fix, so no harm in adding this?

**Comment by juergba** [comment]
@giltayar It depends. Actually I don't see a reason for wrenching this dying horse into our major release v9. In v9 we made a clear statement in favor of Node's ESM, so why should we support the `esm` package? It's just ugly.

So I'm still against this, but evtl. could be convinced if you wrapped your fix in the code refactoring we were talking about.

**Comment by giltayar** [comment]
@juergba the fix is just adding the code `ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING` in:

```js
        if (
          err.code === 'ERR_MODULE_NOT_FOUND' ||
          err.code === 'ERR_UNKNOWN_FILE_EXTENSION' ||
          err.code === 'ERR_UNSUPPORTED_DIR_IMPORT'
        ) {
          return require(file);
        }
```

_really_ simple. 

And, yes, I'll start working on the cleanup I talked about, now that we v9 is out and I'm more confident that it's working correctly without any major ramifications.

I totally agree on `esm`, but if it's just a one line change, I wouldn't deprive people of their use of this package.

**Comment by giltayar** [comment]
OK, I tried. I added the code above, but while trying to write a test that used the `esm` package, I got all sorts of weird unexplainable behaviors. I could spend a few days figuring things out and _maybe_ getting it to work, but as @juergba said, it's not worth it for a package that hasn't been maintained for 2 years.

So, sorry @Yonom, but it seems you'll have to stick to Mocha v8 till you ugprade to native Node.js ESM, or submit a PR (with a test) yourself.
